speechd: refactor and small improvements

pinage404 opened the pull request:
I used to use [Pied](https://search.nixos.org/packages?channel=25.11&show=pied) to have a voice generated by piper

I have [upgraded NixOS from 25.05 to 25.11](https://gitlab.com/pinage404/dotfiles/-/commit/b481b7496e5f724b57f166b0534c690ad5f81b52?page=2#58cb4f58166586c1ed7f076c568d41682df3661c_14_14)

Since speech-dispatcher seems to not work

I was trying to use a more recent version of speech-dispatcher which include piper https://github.com/brailcom/speechd/pull/996

https://github.com/NixOS/nixpkgs/compare/master...pinage404:nixpkgs:speechd-upgrade i let this branch here, in a hope that could give a little help when upgrading after the future release 

With piper, the linker fails, i don't understand why, i tried to add many things in a hope to fix it

While doing all of this thing, i refactored and did some small improvement that can be merge upstream

I tried to have [explicit commits](https://github.com/NixOS/nixpkgs/pull/470797/commits)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pinage404 committed:
speechd: update metadata

meld existing metadata with what is generated by nix-init

pinage404 committed:
speechd: update metadata

meld existing metadata with what is generated by nix-init

pinage404 committed:
speechd: make more explicit python packages

pinage404 committed:
speechd: make more explicit python packages

pinage404 committed:
speechd: add `--with-flite` option

this option was missing

it comes from here:
https://github.com/brailcom/speechd/blob/fbdb6b0bf25e1ec105d5d76e8fcd8a118e97c582/configure.ac#L241

pinage404 committed:
speechd: add `--with-flite` option

this option was missing

it comes from here:
https://github.com/brailcom/speechd/blob/fbdb6b0bf25e1ec105d5d76e8fcd8a118e97c582/configure.ac#L241

pinage404 committed:
speechd: add pipewire option

this option come from here:

https://github.com/brailcom/speechd/blob/fbdb6b0bf25e1ec105d5d76e8fcd8a118e97c582/configure.ac#L496

pinage404 committed:
speechd: add pipewire option

this option come from here:

https://github.com/brailcom/speechd/blob/fbdb6b0bf25e1ec105d5d76e8fcd8a118e97c582/configure.ac#L496

pinage404 committed:
speechd: refactor abort when the substitution fail

pinage404 committed:
speechd: refactor abort when the substitution fail

pinage404 committed:
speechd: actually disable options when disabling with override

pinage404 committed:
speechd: actually disable options when disabling with override

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -143,9 +143,15 @@ stdenv.mkDerivation (finalAttrs: {
+      lib.licenses.gpl2Only
+      lib.licenses.gpl3Only
+      lib.licenses.lgpl21Only
+    ];
```
This does not look right. I see no GPL-3.0 licensed code and according to https://github.com/brailcom/speechd#licensing, the daemon is mainly `GPL-2.0-or-later` and the libraries `LGPL-2.1-or-later`. dotconf is dynamically linked so it should not have an effect on the license of the daemon, IIRC.

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -38,9 +38,11 @@ stdenv.mkDerivation (finalAttrs: {
-  src = fetchurl {
-    url = "https://github.com/brailcom/speechd/releases/download/${finalAttrs.version}/speech-dispatcher-${finalAttrs.version}.tar.gz";
-    sha256 = "sha256-sUpSONKH0tzOTdQrvWbKZfoijn5oNwgmf3s0A297pLQ=";
+  src = fetchFromGitHub {
```
This is already built from source. Generally, I prefer using the official upstream distribution method since that is likely the most tested.

jtojnar commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -70,6 +70,10 @@ stdenv.mkDerivation (finalAttrs: {
     })
   ];
 
+  postPatch = lib.optionalString withPico ''
```
I generally prefer declarative then imperative grouping as described in https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/attribute-ordering.md

The imperative stuff would ideally disappear eventually.

jtojnar reviewed (commented):
Thanks.

Please note that the package is named `speechd` so the commits should start with that as a prefix.

pinage404 commented:
It is not clear for me, sorry

@jtojnar what should i change to make you happy with it ? what could make you approve this PR ?

pinage404 commented:
There are the tags `10.rebuild-linux: 501+` and  `10.rebuild-linux: 501-1000`

Should i change the base of this PR from `master` to `staging` or `staging-next` or `staging-25.11` or `staging-next-25.11` ?

linsui commented:
It should target staging.

pinage404 commented:
I rebased to solve the conflict (i rebased on `master`, then `staging`, i forgot which was the target branch)

WiredMic commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -103,37 +114,31 @@ stdenv.mkDerivation (finalAttrs: {
+    [
+      "--sysconfdir=/etc"
+      # Audio method falls back from left to right.
+      "--with-default-audio-method=\"libao,pulse,pipewire,alsa,oss\""
```
```suggestion
    ''--with-default-audio-method="${
      lib.concatStringsSep "," (
        lib.optional withLibao "libao"
        ++ lib.optional withPulse "pulse"
        ++ lib.optional withAlsa "alsa"
        ++ lib.optional withPipewire "pipewire"
        ++ lib.optional withOss "oss"
      )
    }"''
```
Having a hard coded string of audio output methods means that speechd could use a method, even if it is not enabled.

WiredMic commented on pkgs/by-name/sp/speechd/package.nix:
```suggestion
version = "0.12.1";

__structuredAttrs = true;
strictDeps = true;

src = fetchurl {
url = "https://github.com/brailcom/speechd/releases/download/${finalAttrs.version}/speech-dispatcher-${finalAttrs.version}.tar.gz";
sha256 = "sha256-sUpSONKH0tzOTdQrvWbKZfoijn5oNwgmf3s0A297pLQ=";
```

WiredMic replied:
I have ran and compiled speechd with the newer stricter defaults

WiredMic commented on pkgs/by-name/sp/speechd/package.nix:
```suggestion
  fetchpatch,
  testers,
```
Could it make sense to add a small version test

WiredMic commented on pkgs/by-name/sp/speechd/package.nix:
```diff
@@ -154,8 +159,10 @@ stdenv.mkDerivation (finalAttrs: {
 
   meta = {
```
```suggestion
enableParallelBuilding = true;

passthru = {
  tests = lib.optionalAttrs (!libsOnly) {
    version = testers.testVersion {
      package = finalAttrs.finalPackage;
    };
  };
};

meta = {
```
Could it make sense to add a small version test

WiredMic reviewed (changes requested):
I verified that speechd compiles with `__structuredAttrs = true` and `strictDeps = true`.

I build these native builds
```bash
nix build .#speechd
nix build .#speechd-minimal
nix build --impure --expr '(import <nixpkgs> {}).speechd.override { libsOnly = true; }'
```

I also build one consumer
```bash
nix build .#orca
```

I build speech with a cross build to check that `strictDeps` actually enforced, build platform ≠ host platform
```zsh
nix build .#pkgsCross.gnu64.speechd
```

I build the version test
```bash
# passes on the default build
nix build .#speechd.tests.version

# correctly omitted/guarded under libsOnly, since `bin` is stripped in postInstall
nix build --impure --expr '((import ./. {}).speechd.override { libsOnly = true; }).tests.version'
```

pinage404 committed:
speechd: add testers

pinage404 committed:
speechd: stricter deps

pinage404 committed:
speechd: ensure to use the same option has given arguments

pinage404 committed:
speechd: add which is supported in description

There is many `*-minimal` / `*-full` packages where [NixOS Search](https://search.nixos.org/packages?) don't show any differences

To know the differences, we have to go to the source

I am lazy, so i don't like this and i don't want to add one more

So i added the differences to the description for the minimal package

pinage404 commented:
@WiredMic is there anything that you would like to change to remove the change request ?
